__uk_init_tls.c

```diff
@@ -0,0 +1,200 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <uk/tcb_impl.h>
+#include <elf.h>
+#include <limits.h>
+#include <sys/mman.h>
+#include <string.h>
+#include <stddef.h>
+#include "pthread_impl.h"
+#include "libc.h"
+#include "atomic.h"
+#include "syscall.h"
+
+#include <uk/alloc.h>
+#include <uk/assert.h>
+
+/*
+ * This glue code source is meant to replace `__libc_start_main()`
+ * and tls related stuff.
+ * If the pthread API is used, musl will allocate the TLS and will
+ * provide a pointer to it in the `clone()` syscall. Musl will allocate
+ * in fact a contiguous region which will have the following layout (Please
+ * refer to the `pthread_create.c` file):
+ * map -----------------------------------------------------------------
+ *                     ^
+ *                     |
+ *                     | 4096      GUARD PAGE
+ *                     |
+ *                     v
+ * map + guard ---------------------------------------------------------
+ *                     ^
+ *                     |
+ *                     |
+ *                     |           STACK
+ *                     |
+ *                     |
+ *                     |
+ *                     v
+ * stack ---------------------------------------------------------------
+ *                     ^           ^
+ *                     |           |       TLS SPACE
+ *                     |           v
+ *               new ->|------------------------------------------------
+ *                     |           ^
+ *                     |           |
+ *                     |           |
+ *                     |           | 280   pthread structure
+ *                     |           |
+ *                     v           v
+ * tsd -----------------------------------------------------------------
+ *                     ^
+ *                     | 1024
+ *                     v
+ * map + size ----------------------------------------------------------
+ */
+
+
+void *__uk_copy_tls(unsigned char *mem)
+{
+	pthread_t td;
+	void *tls_area;
+
+	mem -= (uintptr_t)mem & (libc.tls_align-1);
+	tls_area = mem;
+	ukarch_tls_area_init(tls_area);
+
+	td = (pthread_t) ukarch_tls_tlsp(tls_area);
+	td->dtv = td->dtv_copy = tls_area;
+
+	return td;
+}
+
+static int __uk_init_tp(void *p)
+{
+	pthread_t td = p;
+
+	td->self = td;
+	/*
+	 * Set the `$fs` register for the current thread.
+	 * In the original code of musl this will use an `arch_prtcl`
+	 * syscall to fill the `$fs` register.
+	 */
+	ukplat_tlsp_set(TP_ADJ(p));
```

Comment:
```suggestion
	ukplat_tlsp_set((unsigned long) TP_ADJ(p));
```

To suppress a warning.

Reply:
Thank you for your comment!